### PR TITLE
Add structured logging (json) to whole server code base: file/error/console logs and stats (except ClientAPI)

### PR DESCRIPTION
**Pull request by shaan1337**
- This PR adds structured logging to the whole server code base: file/error/console logs and stats (no support for ClientAPI yet) preserving backward compatibility
- Add option --structured-log/StructuredLog to enable structured logging on startup
- Upgrade NLog to version 4.5.6
- Improve logging performance by adding `keepFileOpen="true"`

Co-authored-by: Avish Cheetaram <avishcheetaram.ai@gmail.com>
Co-authored-by: Lougarou <yogisawesome@gmail.com>
Co-authored-by: Shaan Nobee <sniper111@gmail.com>

### CODE REVIEW

File: src/EventStore.Common/Log/log.config [hunk 1]
```diff
@@ -11,36 +11,93 @@
...
     </target>
-    
-    <target name="coloredConsoleLog" xsi:type="ColoredConsole" layout="${ConsoleLayout}" useDefaultRowHighlightingRules="true">
-      <highlight-row condition="level == LogLevel.Info" foregroundColor="Green" backgroundColor="NoChange"/>
+    <target name="errorFileLogStructured" xsi:type="FilteringWrapper" condition="length('${exception}')>0 || level==LogLevel.Error || level==LogLevel.Fatal">
+      <target xsi:type="File" fileName="${LogFileNameBase}-err.log" keepFileOpen="true" openFileCacheTimeout="30">
+          <layout xsi:type="JsonLayout">
+            <attribute name="PID" layout="${processid}" />
+            <attribute name="ThreadID" layout="${threadid}" />
+            <attribute name="Date" layout="${date:universalTime=true:format=yyyy\.MM\.dd HH\:mm\:ss\.fff}" />
```
Comment (shaan1337) [inline src/EventStore.Common/Log/log.config]: format needs to be fixed to `yyyy\-MM\-ddTHH\:mm\:ss\.ffffffZ`

**Comment by snakefoot** [comment]
@shaan1337 Any reason why you are not referencing the NLog-nuget-package in `EventStore.Core.Tests.csproj`:

```
<Reference Include="NLog">
<HintPath>..\libs\NLog.dll</HintPath>
</Reference>
```

Along with removing this file from the repository:

>  src/libs/NLog.dll

You can also optimize the `log.config` by adding timeToSleepBetweenBatches to the default-wrapper (Supported by NLog ver. 4.3.11 or later): 

`<default-wrapper xsi:type="AsyncWrapper" overflowAction="Grow" timeToSleepBetweenBatches="0" />`

It will optimize the performance of the AsyncWrapper, and also make the system more "silent" when idle (reducing the number of timer events)

File: src/EventStore.Common/Log/log.config [hunk 2]
```diff
@@ -5,42 +5,99 @@
...
-  <variable name="ConsoleLayout" value="[${processid:padCharacter=0:padding=5},${threadid:padCharacter=0:padding=2},${date:universalTime=true:format=HH\:mm\:ss\.fff}] ${message}${onexception:${newline}EXCEPTION OCCURRED:${newline}${exception:format=message}}"/>
+  <variable name="VerboseLayout" value="[PID:${processid:padCharacter=0:padding=5}:${threadid:padCharacter=0:padding=3} ${date:universalTime=true:format=yyyy\.MM\.dd HH\:mm\:ss\.fff} ${level:padding=-5:uppercase=true} ${logger:padding=-20:fixedLength=true}] ${message}${onexception:${newline}${literal:text=EXCEPTION OCCURRED}${newline}${exception:format=tostring:innerFormat=tostring:maxInnerExceptionLevel=20}}"/>
+  <variable name="ConsoleLayout" value="[${processid:padCharacter=0:padding=5},${threadid:padCharacter=0:padding=2},${date:universalTime=true:format=HH\:mm\:ss\.fff}] ${message}${onexception:${newline}${literal:text=EXCEPTION OCCURRED}${newline}${exception:format=message}}"/>
   <variable name="LogFileNameBase" value="${logsdir}/${shortdate:universalTime=true}/${environment:variable=EVENTSTORE_INT-COMPONENT-NAME}"/>
   
   <targets>
     <default-wrapper xsi:type="AsyncWrapper" overflowAction="Grow" />
```
Comment (shaan1337) [inline src/EventStore.Common/Log/log.config]: Add `timeToSleepBetweenBatches="0" ` as per recommendations of @snakefoot

**Comment by shaan1337** [comment]
thanks for the tip @snakefoot, we'll add this. we have only a few external dependencies, so just including the DLLs simplifies the build/release process

**Comment by gregoryyoung** [comment]
My guess is that this is historical.

File: src/EventStore.Core/Services/Gossip/GossipServiceBase.cs [hunk 3]
```diff
@@ -183,11 +184,11 @@ public void Handle(GossipMessage.GossipSendFailed message)
 
             if (CurrentMaster != null && node.InstanceId == CurrentMaster.InstanceId)
             {
-                Log.Trace("Looks like master [{0}, {1:B}] is DEAD (Gossip send failed), though we wait for TCP to decide.",
+                Log.Trace("Looks like master [{recipient}, {instanceId:B}] is DEAD (Gossip send failed), though we wait for TCP to decide.",
```
Comment (shaan1337) [inline src/EventStore.Core/Services/Gossip/GossipServiceBase.cs]: rename `recipient`

File: src/EventStore.ClientAPI/Transport.Tcp/TcpPackageConnection.cs [hunk 4]
```diff
@@ -76,6 +76,7 @@ internal class TcpPackageConnection
                     log.Debug("TcpPackageConnection: connection to [{0}, L{1}, {2:B}] failed. Error: {3}.", conn.RemoteEndPoint, conn.LocalEndPoint, connectionId, error);
                     if (connectionClosed != null)
                         connectionClosed(this, error);
+                    log.Debug("TcpPackageConnection: connection to [{0}, L{localEndPoint}, {connectionId:B}] failed. Error: {e}.", conn.RemoteEndPoint, conn.LocalEndPoint, connectionId, error);
```
Comment (shaan1337) [inline src/EventStore.ClientAPI/Transport.Tcp/TcpPackageConnection.cs]: remove this line, seems like a wrong merge

File: src/EventStore.Common/Log/NLogger.cs [hunk 5]
```diff
@@ -64,31 +70,32 @@ public void Warn(string format, params object[] args)
         public void Trace(string format, params object[] args)
         {
             _logger.Trace(format, args);
+
```
Comment (shaan1337) [inline src/EventStore.Common/Log/NLogger.cs]: remove extra line

File: src/EventStore.Common/Log/log.config [hunk 6]
```diff
@@ -5,42 +5,99 @@
...
+        <layout xsi:type="JsonLayout">
+            <attribute name="PID" layout="${processid}" />
+            <attribute name="ThreadID" layout="${threadid}" />
+            <attribute name="Date" layout="${date:universalTime=true:format=yyyy\-MM\-ddTHH\:mm\:ss\.ffffffZ}" />
+            <attribute name="Level" layout="${level}" />
+            <attribute name="Logger" layout="${logger}" />
+            <attribute name="Message" layout="${message:raw=true}" />
+            <attribute name="Exception" layout="${exception:format=@}" encode="false"/>
+            <attribute name="EventProperties" encode="false" >
+            <layout type='JsonLayout' includeAllProperties="true"/>
```
Comment (shaan1337) [inline src/EventStore.Common/Log/log.config]: verify this line

**Comment by shaan1337** [comment]
Manual review completed - several commits pushed.

**Comment by snakefoot** [comment]
It is nice of you to copy the source code of NLog and taking it to your bosom, but if you only need to remove quotes around strings, then you can also do this:

```c#
        class OverrideValueFormatter : IValueFormatter
        {
            private IValueFormatter _originalFormatter;

            public bool IsStructured { get; set; }

            public OverrideValueFormatter(IValueFormatter originalFormatter)
            {
                _originalFormatter = originalFormatter;
            }

            public bool FormatValue(object value, string format, NLog.MessageTemplates.CaptureType captureType, IFormatProvider formatProvider, System.Text.StringBuilder builder)
            {
                if (!IsStructured && captureType == NLog.MessageTemplates.CaptureType.Normal)
                {
                    switch (Convert.GetTypeCode(value))
                    {
                        case TypeCode.String:
                            {
                                builder.Append((string)value);
                                return true;
                            }
                        case TypeCode.Char:
                            {
                                builder.Append((char)value);
                                return true;
                            }
                        case TypeCode.Empty:
                            return true;  // null becomes empty string
                    }
                }

                return _originalFormatter.FormatValue(value, format, captureType, formatProvider, builder);
            }
        }
```

And then use it like this:

```c#
            var oldFormatter = NLog.Config.ConfigurationItemFactory.Default.ValueFormatter;
            var newFormatter = new OverrideValueFormatter(oldFormatter);
            newFormatter.IsStructured = false;
            NLog.Config.ConfigurationItemFactory.Default.ValueFormatter = newFormatter;
```

Then you can remove these files:

```
+    <Compile Include="Log\ESValueFormatter.cs" />
+    <Compile Include="Log\MruCache.cs" />
+    <Compile Include="Log\SingleItemOptimizedHashSet.cs" />
```

**Comment by shaan1337** [comment]
hey @snakefoot thanks that's perfect, we didn't find an easier way earlier

**Comment by shaan1337** [comment]
Additional logging changes made after rebase to `master`

**Pull request merged**